internal/app/machined/pkg/controllers/siderolink/userspace.go

```diff
@@ -0,0 +1,200 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package siderolink
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"sync/atomic"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/optional"
+	"github.com/siderolabs/siderolink/pkg/wgtunnel"
+	"github.com/siderolabs/siderolink/pkg/wgtunnel/wgbind"
+	"github.com/siderolabs/siderolink/pkg/wgtunnel/wggrpc"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/siderolabs/talos/internal/pkg/ctxutil"
+	"github.com/siderolabs/talos/internal/pkg/endpoint"
+	"github.com/siderolabs/talos/pkg/machinery/resources/config"
+	"github.com/siderolabs/talos/pkg/machinery/resources/siderolink"
+)
+
+// UserspaceWireguardController imlements a controller that manages a Wireguard over GRPC tunnel in userspace.
+type UserspaceWireguardController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *UserspaceWireguardController) Name() string {
+	return "siderolink.UserspaceWireguardController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *UserspaceWireguardController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      siderolink.TunnelType,
+			ID:        optional.Some(siderolink.TunnelID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *UserspaceWireguardController) Outputs() []controller.Output {
+	return []controller.Output{}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *UserspaceWireguardController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	eg, ctx := errgroup.WithContext(ctx)
+
+	logger.WithOptions()
+
+	var (
+		tunnelDevice atomic.Pointer[wgtunnel.TunnelDevice]
+		tunnelRelay  atomic.Pointer[tunnelProps]
+	)
+
+	defer func() {
+		tunnelRelay.Load().Relay().Close()
+		tunnelDevice.Load().Close()
+	}()
+
+	qp := wgbind.NewQueuePair(100, 100)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctxutil.Cause(ctx)
+		case <-r.EventCh():
+		}
+
+		res, err := safe.ReaderGetByID[*siderolink.Tunnel](ctx, r, siderolink.TunnelID)
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				tunnelRelay.Load().Relay().Close()
+				tunnelDevice.Load().Close()
+
+				continue
+			}
+
+			return fmt.Errorf("failed to read link spec: %w", err)
+		}
+
+		if td := tunnelDevice.Load(); td.IsClosed() {
+			td.Close()
+
+			td, err = wgtunnel.NewTunnelDevice(res.TypedSpec().IFaceName, res.TypedSpec().MTU, qp, logger)
+			if err != nil {
+				return fmt.Errorf("failed to create tunnel device: %w", err)
+			}
+
+			tunnelDevice.Store(td)
+
+			logger.Info("wg over grpc tunnel device created", zap.String("interface", res.TypedSpec().IFaceName))
+
+			eg.Go(func() error {
+				logger.Info("running tunnel device")
```

Comment:
should be `.Debug` ? why should user care about it?

Reply:
Fixed!